### PR TITLE
Event size validation for Wodles

### PR DESCRIPTION
**Pull request by RamosFe**
|Related issue|
|---|
| [#17689](https://github.com/wazuh/wazuh/issues/17689) |

## Description

Added validations for the size of the events sent to the Analysisd queue.

### CODE REVIEW

File: wodles/aws/aws_s3.py [hunk 1]
```diff
@@ -513,8 +511,15 @@ def send_msg(self, msg, dump_json=True):
...
             s.connect(self.wazuh_queue)
-            s.send("{header}{msg}".format(header=self.msg_header,
-                                          msg=json_msg if dump_json else msg).encode())
+
+            encoded_msg = "{header}{msg}".format(header=self.msg_header,
+                                                 msg=json_msg if dump_json else msg).encode()
+
+            # Logs warning if event is bigger than max size
+            if len(encoded_msg) > MAX_EVENT_SIZE:
+                debug(f"WARNING: : Event size exceeds the maximum allowed limit of {MAX_EVENT_SIZE} bytes.")
```
Comment (fdalmaup) [inline wodles/aws/aws_s3.py:520]: The `debug` function requires the `msg_level` argument

**Review by fdalmaup — CHANGES_REQUESTED** [review]
The modifications seem right but there is no manual testing evidence in the PR nor the issue. Please add it for at least one of the integrations.

**Comment by RamosFe** [comment]
## Update

Created a new file with a size bigger than the limit set in the new validations:

image

This file has only 1 line with a string bigger than the max limit (Because the Azure Storage Wodle sends the bia socket file line by line, the file only has 1 line to send the string with the desired size and raise the warning).
Script execution:

```
# ./azure-logs --storage --storage_auth_path /var/ossec/wodles/azure/azure_credentials --container bigblob --reparse
2023/07/21 01:51:46 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/07/21 01:51:46 azure: ERROR: ERROR: Message too long to send to Wazuh.  Skipping message...
2023/07/21 01:51:47 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/07/21 01:51:47 azure: ERROR: ERROR: Message too long to send to Wazuh.  Skipping message...
2023/07/21 01:51:47 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
```

File: wodles/aws/aws_s3.py [hunk 2]
```diff
@@ -513,8 +511,15 @@ def send_msg(self, msg, dump_json=True):
...
             s = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
             s.connect(self.wazuh_queue)
-            s.send("{header}{msg}".format(header=self.msg_header,
-                                          msg=json_msg if dump_json else msg).encode())
+
+            encoded_msg = "{header}{msg}".format(header=self.msg_header,
+                                                 msg=json_msg if dump_json else msg).encode()
+
+            # Logs warning if event is bigger than max size
+            if len(encoded_msg) > MAX_EVENT_SIZE:
```
Comment (fdalmaup) [inline wodles/aws/aws_s3.py:519]: The AWS module imports the entire utils file. Therefore, this line will throw an error.
```suggestion
            if len(encoded_msg) > utils.MAX_EVENT_SIZE:
```

File: wodles/azure/azure-logs.py [hunk 3]
```diff
@@ -828,9 +828,16 @@ def send_message(message: str):
         The message body to send to analysisd.
     """
     s = socket(AF_UNIX, SOCK_DGRAM)
+
+    encoded_msg = f'{SOCKET_HEADER}{message}'.encode(errors='replace')
+
+    # Logs warning if event is bigger than max size
+    if len(encoded_msg) > MAX_EVENT_SIZE:
+        logging.warning(f"WARNING: Event size exceeds the maximum allowed limit of {MAX_EVENT_SIZE} bytes.")
```
Comment (fdalmaup) [inline wodles/azure/azure-logs.py:836]: You may remove `WARNING` from the log message since the logger already adds it.

**Review by fdalmaup — CHANGES_REQUESTED** [review]

File: wodles/aws/aws_s3.py (hunk 2 above)
Comment (RamosFe) [reply wodles/aws/aws_s3.py]: Fixed in commit 113469afbaff8582af4db430587402c60c697277:

https://github.com/wazuh/wazuh/blob/113469afbaff8582af4db430587402c60c697277/wodles/aws/aws_s3.py#L519-L520

File: wodles/azure/azure-logs.py (hunk 3 above)
Comment (RamosFe) [reply wodles/azure/azure-logs.py:836]: Fixed in commit 113469afbaff8582af4db430587402c60c697277:

https://github.com/wazuh/wazuh/blob/113469afbaff8582af4db430587402c60c697277/wodles/aws/aws_s3.py#L519-L520

**Review by fdalmaup — APPROVED** [review]
LGTM!

**Review by Selutario — COMMENTED** [review]
In the update https://github.com/wazuh/wazuh/pull/17932#issuecomment-1645635392 I see these two errors:
```
2023/07/21 01:51:46 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/07/21 01:51:46 azure: ERROR: ERROR: Message too long to send to Wazuh.  Skipping message...
```

The first one is the new warning you added and the second one is the error shown when the response code is `90`:
```
elif e.errno == 90:
    logging.error("ERROR: Message too long to send to Wazuh.  Skipping message...")
``` 

However, according to the [issue description](https://github.com/wazuh/wazuh/issues/17689#issue-1779763054), this error is only supposed to appear when the buffer size of `/var/ossec/queue/sockets/queue` is exceeded, which is 212992B. 

Since the file you uploaded is 93 KiB, far from the size of the buffer, could you verify from what event size the response code 90 is being obtained? If it is returned for events of 65 KiB or bigger, this development would not be necessary.

**Comment by RamosFe** [comment]
Added extra error messages only for debugging this case:
### Debugging
```
    try:
        s.connect(ANALYSISD)
        logging.warning(f"WARNING: Event size {len(encoded_msg)}")
        s.send(encoded_msg)
    except for socket_error as e:
        if e.errno == 111:
            logging.error("ERROR: Wazuh must be running.")
            sys.exit(1)
        elif e.errno == 90:
            logging.error("ERROR: Message too long to send to Wazuh.  Skipping message...")
            send_buffer_size = s.getsockopt(SOL_SOCKET, SO_SNDBUF)
            recv_buffer_size = s.getsockopt(SOL_SOCKET, SO_RCVBUF)
            logging.error(f"ERROR: Wazuh send {send_buffer_size} size and recv {recv_buffer_size}")
            logging.error(e)
        else:
            logging.error(f"ERROR: Error sending message to wazuh: {e}")
            sys.exit(1)
    finally:
        s.close()
```
### Result
There are 3 files in the `bigblob` container with the following sizes: 500mb, 5mb, and 93KB. For the first 2 files, the WARNING and ERROR are logged, but for the last file, only the WARNING is logged.
I also added a log message for getting the size of the `send` and `recv` buffers. These two values are always 212992B.

```
./azure-logs --storage --storage_auth_path /var/ossec/wodles/azure/azure_credentials --container bigblob --reparse
2023/08/16 06:44:40 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/08/16 06:44:40 azure: WARNING: WARNING: Event size 524288034
2023/08/16 06:44:40 azure: ERROR: ERROR: Message too long to send to Wazuh.  Skipping message...
2023/08/16 06:44:40 azure: ERROR: ERROR: Wazuh send 212992 size and recv 212992
2023/08/16 06:44:40 azure: ERROR: [Errno 90] Message too long
2023/08/16 06:44:41 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/08/16 06:44:41 azure: WARNING: WARNING: Event size 5242914
2023/08/16 06:44:41 azure: ERROR: ERROR: Message too long to send to Wazuh.  Skipping message...
2023/08/16 06:44:41 azure: ERROR: ERROR: Wazuh send 212992 size and recv 212992
2023/08/16 06:44:41 azure: ERROR: [Errno 90] Message too long
2023/08/16 06:44:41 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/08/16 06:44:41 azure: WARNING: WARNING: Event size 95569
```

**Review by Selutario — CHANGES_REQUESTED** [review]
Fix conflict in aws wodle: 
image

File: wodles/aws/aws_s3.py [hunk 4]
```diff
@@ -540,8 +540,20 @@ def send_msg(self, msg, dump_json=True):
...
+
+            encoded_msg = "{header}{msg}".format(header=self.msg_header,
+                                                 msg=json_msg if dump_json else msg).encode()
+
+            # Logs warning if event is bigger than max size
+            if len(encoded_msg) > utils.MAX_EVENT_SIZE:
+                debug(f"Event size exceeds the maximum allowed limit of {utils.MAX_EVENT_SIZE} bytes.", 1)
+
+            s.send(encoded_msg)
+>>>>>>> 6831ecae28 (Added msg_level in aws_s3.py.)
```
[the remark this answers is not in the document]
Comment (RamosFe) [reply wodles/aws/aws_s3.py]: Fixed in e7a86830c7259c041e2ad24e904b2f2db717851e.

**Review by Selutario — APPROVED** [review]

**Pull request merged**